bundles/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/internal/handler/PentairControllerHandler.java

```diff
@@ -0,0 +1,1186 @@
+                                   // but it requires to be in sync and up-to-date
+    private long lastScheduleTypeWrite;
+
+    private static final int CACHE_EXPIRY = (int) TimeUnit.SECONDS.toMillis(10);
+    private static final int CACHE_EXPIRY_LONG = (int) TimeUnit.MINUTES.toMillis(10);
+
+    protected ExpiringCache<PentairControllerStatus> controllerStatusCache = new ExpiringCache<>(CACHE_EXPIRY);
+    protected ExpiringCache<PentairHeatStatus> heatStatusCache = new ExpiringCache<>(CACHE_EXPIRY);
+
+    private int majorrev, minorrev;
+
+    protected ExpiringCache<PentairControllerCircuit>[] circuitsCache = new ExpiringCache[NUM_CIRCUITS];
+    protected ExpiringCache<PentairControllerSchedule>[] schedulesCache = new ExpiringCache[NUM_SCHEDULES];
+
+    @Nullable
+    protected LightMode lightMode;
+
+    public PentairControllerHandler(Thing thing) {
+        super(thing);
+
+        for (int i = 0; i < NUM_SCHEDULES; i++) {
+            schedulesCache[i] = new ExpiringCache<>(CACHE_EXPIRY_LONG);
+        }
+
+        for (int i = 0; i < NUM_CIRCUITS; i++) {
+            circuitsCache[i] = new ExpiringCache<>(CACHE_EXPIRY_LONG);
+        }
+    }
+
+    @Override
+    public void getConfiguration() {
+        PentairControllerHandlerConfig config = getConfigAs(PentairControllerHandlerConfig.class);
+
+        this.id = config.id;
+        this.syncTime = config.synctime;
+    }
+
+    @Override
+    public void goOnline() {
+        if (onlineController != null) {
+            logger.debug("Another controller is already configured");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Another controller is already configured.");
+        } else {
+            super.goOnline();
+        }
+    }
+
+    @Override
+    public void finishOnline() {
+        super.finishOnline();
+
+        // setup timer to sync time
+        Runnable runnable = new Runnable() {
+
+            @Override
+            public void run() {
+                boolean synctime = ((boolean) getConfig().get("synctime"));
+                if (synctime) {
+                    logger.debug("Synchronizing System Time with Pentair controller");
+                    Calendar now = Calendar.getInstance();
+                    setClockSettings(now.get(Calendar.HOUR_OF_DAY), now.get(Calendar.MINUTE),
+                            now.get(Calendar.DAY_OF_WEEK), now.get(Calendar.DAY_OF_MONTH), now.get(Calendar.MONTH) + 1,
+                            now.get(Calendar.YEAR) - 2000);
+                }
+            }
+        };
+
+        // setup syncTimeJob to run once a day, initial time to sync is 3 minutes after controller goes online. This is
+        // to prevent collision with main thread queries on initial startup
+        syncTimeJob = scheduler.scheduleWithFixedDelay(runnable, 3, 24 * 60 * 60, TimeUnit.MINUTES);
+
+        Runnable queryinfo = new Runnable() {
```

Comment:
please use lambda here. Although a better suggestion is to move this logic into its own method and call that with a lambda instead. That suggestion also applies to my earlier comment.